katz/src/main/kotlin/katz/effects/IO.kt

```diff
@@ -0,0 +1,220 @@
+package katz
+
+import katz.effects.internal.AndThen
+import katz.effects.internal.Platform
+import katz.effects.internal.Platform.unsafeResync
+
+sealed class IO<out A> {
+    abstract fun <B> map(f: (A) -> B): IO<B>
+
+    fun <B> flatMap(f: (A) -> IO<B>): IO<B> =
+            flatMapTotal(
+                    AndThen { a: A ->
+                        try {
+                            f(a)
+                        } catch (error: Throwable) {
```

Comment:
Do we want to catch nay Throwable or just those that are non fatal? https://github.com/scala/scala/blob/v2.9.3/src/library/scala/util/control/NonFatal.scala#L31